javalib/src/main/scala/java/time/Duration.scala

```diff
@@ -0,0 +1,240 @@
+
+  def negated(): Duration = Duration(totalNanos * -1)
+
+  def abs(): Duration = if (isNegative()) negated() else this
+
+  def addTo(temporal: Temporal): Temporal =
+    temporal.plus(seconds, ChronoUnit.SECONDS).plus(nanos, ChronoUnit.NANOS)
+
+  def subtractFrom(temporal: Temporal): Temporal =
+    temporal.minus(seconds, ChronoUnit.SECONDS).minus(nanos, ChronoUnit.NANOS)
+
+  def toDays(): Long = seconds / SECONDS_IN_DAY
+
+  def toHours(): Long = seconds / SECONDS_IN_HOUR
+
+  def toMinutes(): Long = seconds / SECONDS_IN_MINUTE
+
+  def toMillis(): Long = {
+    val (quot, rem) = totalNanos /% NANOS_IN_MILLI
+    val millis = if (rem < 0) quot - 1 else quot
+    // Reference implementation throws if millis < -9223372036854775000!
+    if (millis >= -9223372036854775000L && millis <= Long.MaxValue)
+      millis.toLong
+    else
+      throw new ArithmeticException("Long overflow")
+  }
+
+  def toNanos(): Long = {
+    // Reference implementation throws if nanos < -9223372036000000000!
+    if (totalNanos >= -9223372036000000000L && totalNanos <= Long.MaxValue)
+      totalNanos.toLong
+    else
+      throw new ArithmeticException("Long overflow")
+  }
+
+  def compareTo(that: Duration) = {
+    val secCmp = seconds.compareTo(that.getSeconds)
+    if (secCmp == 0) nanos.compareTo(that.getNano)
+    else secCmp
+  }
+
+  override def equals(that: Any): Boolean = that match {
+    case that: Duration =>
+      seconds == that.getSeconds && nanos == that.getNano
+
+    case _ => false
+  }
+
+  override def hashCode(): Int = 31 * seconds.hashCode + nanos
+
+  override def toString(): String = {
+    val nanos1 = if (seconds >= 0 || nanos == 0) nanos else NANOS_IN_SECOND - nanos
+    val secs = if (seconds >= 0 || nanos == 0) seconds else seconds + 1
+    val mins = secs / 60
+    val secsOfMin = secs % 60
+    val hours = mins / 60
+    val minsOfHour = mins % 60
+    val hourPart = if (hours == 0) "" else hours.toString + "H"
+    val minPart = if (minsOfHour == 0) "" else minsOfHour.toString + "M"
+    val decimals = f"$nanos1%09d".reverse.dropWhile(_ == '0').reverse
+    val decimalPart = if (decimals.isEmpty) "" else "." + decimals
+    val secsPart = secsOfMin match {
+      case 0 if seconds != 0 && nanos == 0 => ""
+      case 0 if seconds < 0                => "-0" + decimalPart + "S"
+      case n                               => n.toString + decimalPart + "S"
+    }
+    "PT" + hourPart + minPart + secsPart
+  }
+}
+
+object Duration {
+  import Constants._
+
+  private def apply(nanos: BigInt): Duration = {
+    val (quot, rem) = nanos /% NANOS_IN_SECOND
+    val seconds = if (rem < 0) quot - 1 else quot
+    val nanosOfSeconds = if (rem < 0) rem + NANOS_IN_SECOND else rem
+    if (seconds.isValidLong) new Duration(seconds.toLong, nanosOfSeconds.toInt)
+    else throw new ArithmeticException("Long overflow")
+  }
+
+  val ZERO = new Duration(0, 0)
+
+  def ofDays(days: Long): Duration = Duration(BigInt(days) * NANOS_IN_DAY)
```

Comment:
You could implement all the `ofXXX` methods without the need of an intermediate `BigInt` by using `of(amount, unit)`. 
```scala
def ofDays(days: Long): Duration = of(days, ChronoUnit.DAYS)
def ofHours(hours: Long): Duration = of(days, ChronoUnit.HOURES)
...
```
Of course currently `of` is using `multipliedBy` which uses `BigInt`s, but tat should also change if we want a decent performance.